File: gitdot-web/app/dal/repository.ts
import "server-only";

import {
  type CreateRepositoryRequest,
  type GetRepositoryBlobRequest,
  type GetRepositoryBlobsRequest,
  type GetRepositoryCommitsRequest,
  type GetRepositoryFileCommitsRequest,
  type GetRepositoryPathsRequest,
  type GetRepositoryPreviewRequest,
  RepositoryBlobResource,
  RepositoryBlobsResource,
  RepositoryCommitResource,
  RepositoryCommitsResource,
  RepositoryPathsResource,
  RepositoryPreviewResource,
  RepositoryResource,
} from "gitdot-api";
import { toQueryString } from "@/util";
import {
  authDelete,
  authFetch,
  authPost,
  GITDOT_SERVER_URL,
  handleResponse,
} from "./util";

export async function createRepository(
  owner: string,
  repo: string,
  request: CreateRepositoryRequest,
): Promise<RepositoryResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
    request,
  );

  return await handleResponse(response, RepositoryResource);
}

export async function getRepositoryBlob(
  owner: string,
  repo: string,
  query: GetRepositoryBlobRequest,
): Promise<RepositoryBlobResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/blob?${queryString}`,
  );

  return await handleResponse(response, RepositoryBlobResource);
}

export async function getRepositoryCommits(
  owner: string,
  repo: string,
  query?: GetRepositoryCommitsRequest,
): Promise<RepositoryCommitsResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commits?${queryString}`,
  );

  return await handleResponse(response, RepositoryCommitsResource);
}

export async function getRepositoryFileCommits(
  owner: string,
  repo: string,
  query: GetRepositoryFileCommitsRequest,
): Promise<RepositoryCommitsResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/file/commits?${queryString}`,
  );

  return await handleResponse(response, RepositoryCommitsResource);
}

export async function getRepositoryCommit(
  owner: string,
  repo: string,
  sha: string,
): Promise<RepositoryCommitResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commits/${sha}`,
  );

  return await handleResponse(response, RepositoryCommitResource);
}

export async function getRepositoryPreview(
  owner: string,
  repo: string,
  query?: GetRepositoryPreviewRequest,
): Promise<RepositoryPreviewResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/preview?${queryString}`,
  );

  return await handleResponse(response, RepositoryPreviewResource);
}

export async function getRepositoryPaths(
  owner: string,
  repo: string,
  query?: GetRepositoryPathsRequest,
): Promise<RepositoryPathsResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/paths?${queryString}`,
  );
  return await handleResponse(response, RepositoryPathsResource);
}

export async function getRepositoryBlobs(
  owner: string,
  repo: string,
  request: GetRepositoryBlobsRequest,
): Promise<RepositoryBlobsResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/blobs`,
    request,
  );
  return await handleResponse(response, RepositoryBlobsResource);
}

export async function deleteRepository(
  owner: string,
  repo: string,
): Promise<void> {
  const response = await authDelete(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
  );
  if (!response.ok) {
    let message = response.statusText;
    try {
      const body = await response.json();
      if (typeof body?.message === "string") message = body.message;
    } catch {}
    throw new Error(message);
  }
}

File: gitdot-server/src/handler/repository.rs
mod create_repository;
mod delete_repository;
mod get_repository_blob;
mod get_repository_blobs;
mod get_repository_commit;
mod get_repository_commit_diff;
mod get_repository_commits;
mod get_repository_paths;
mod get_repository_resources;
mod get_repository_settings;
mod update_repository_settings;

use axum::{
    Router,
    routing::{get, post},
};

use crate::app::AppState;

use create_repository::create_repository;
use delete_repository::delete_repository;
use get_repository_blob::get_repository_blob;
use get_repository_blobs::get_repository_blobs;
use get_repository_commit::get_repository_commit;
use get_repository_commit_diff::get_repository_commit_diff;
use get_repository_commits::get_repository_commits;
use get_repository_paths::get_repository_paths;
use get_repository_resources::get_repository_resources;
use get_repository_settings::get_repository_settings;
use update_repository_settings::update_repository_settings;

pub fn create_repository_router() -> Router<AppState> {
    Router::new()
        .route(
            "/repository/{owner}/{repo}",
            post(create_repository).delete(delete_repository),
        )
        .route("/repository/{owner}/{repo}/blob", get(get_repository_blob))
        .route(
            "/repository/{owner}/{repo}/blobs",
            post(get_repository_blobs),
        )
        .route(
            "/repository/{owner}/{repo}/paths",
            get(get_repository_paths),
        )
        .route(
            "/repository/{owner}/{repo}/commits",
            get(get_repository_commits),
        )
        .route(
            "/repository/{owner}/{repo}/commits/{sha}",
            get(get_repository_commit),
        )
        .route(
            "/repository/{owner}/{repo}/commits/{sha}/diff",
            get(get_repository_commit_diff),
        )
        .route(
            "/repository/{owner}/{repo}/settings",
            get(get_repository_settings).patch(update_repository_settings),
        )
        .route(
            "/repository/{owner}/{repo}/resources",
            post(get_repository_resources),
        )
}
